include/apq/impl/async_connect.h

```diff
@@ -0,0 +1,172 @@
+    }
+
+    boost::system::error_code ec;
+    get_socket(conn).assign(new_fd, ec);
+
+    return make_error_code(ec, "assign socket failed");
+}
+} // namespace pq
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code start_connection(T& conn, const std::string& conninfo) {
+    using pq::pq_start_connection;
+    return pq_start_connection(unwrap_connection(conn), conninfo);
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code assign_socket(T& conn) {
+    using pq::pq_assign_socket;
+    return pq_assign_socket(unwrap_connection(conn));
+}
+
+template <typename T, typename Handler, typename = Require<Connectiable<T>>>
+inline void write_poll(T& conn, Handler&& h) {
+    using pq::pq_write_poll;
+    pq_write_poll(unwrap_connection(conn), std::forward<Handler>(h));
+}
+
+template <typename T, typename Handler, typename = Require<Connectiable<T>>>
+inline void read_poll(T& conn, Handler&& h) {
+    using pq::pq_read_poll;
+    pq_read_poll(unwrap_connection(conn), std::forward<Handler>(h));
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline decltype(auto) connect_poll(T& conn) {
+    using pq::pq_connect_poll;
+    return pq_connect_poll(unwrap_connection(conn));
+}
+
+/**
+* Asynchronous connection operation
+*/
+template <typename Handler, typename Connection>
+struct async_connect_op {
+    static_assert(Connectiable<Connection>,
+        "Connection type does not meet Connectiable requirements");
+
+    Connection& conn_;
+    Handler handler_;
+
+    void done(error_code ec = error_code{}) {
+        get_io_context(conn_)
+            .post(detail::bind(std::move(handler_), std::move(ec)));
+    }
+
+    void perform(const std::string& conninfo) {
+        error_code ec = start_connection(conn_, conninfo);
+        if (ec) {
+            return done(ec);
+        }
+
+        if (connection_bad(conn_)) {
+            std::ostringstream msg;
+            msg << "Connection is bad";
+            const auto error = error_message(conn_);
+            if (!error.empty()) {
+                msg << ": " << error;
+            }
+            msg << " (" << conninfo << ")";
+            return done(make_error_code(error::pq_connection_status_bad,
+                msg.str()));
+        }
+
+        ec = assign_socket(conn_);
+        if (ec) {
+            return done(ec);
+        }
+
+        return write_poll(conn_, std::move(*this));
+    }
+
+    void operator () (boost::system::error_code ec, std::size_t = 0) {
+        if (ec) {
+            return done(make_error_code(ec, "error while connection polling"));
+        }
+
+        switch (connect_poll(conn_)) {
+        case PGRES_POLLING_OK:
+            return done();
+
+        case PGRES_POLLING_WRITING:
+            return write_poll(conn_, std::move(*this));
+
+        case PGRES_POLLING_READING:
+            return read_poll(conn_, std::move(*this));
+
+        case PGRES_POLLING_FAILED:
+        default:
```

Comment:
Why do you ignore other values?

Reply:
They are not ignored - we didn't expect them at all. So any other value is unexpected and leads to the end of the procedure with an error. Of course, we can remove PGRES_POLLING_FAILED and nothing will change now. Or we can add one more error code for PGRES_POLLING_FAILED only.